org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/embedder/MavenImpl.java

```diff
@@ -174,7 +174,7 @@
 
 
 @Component(service = {IMaven.class, IMavenConfigurationChangeListener.class})
-public class MavenImpl implements IMaven, IMavenConfigurationChangeListener {
+public class MavenImpl implements IMaven, IMavenConfigurationChangeListener, Cloneable {
```

Comment:
I don't think that `Cloneable` is a good choice here and a (private) copy constructor would be much more suitable.

Reply:
What's the particular issue with Cloneable?

Reply:
I found this rather descriptive: https://www.vojtechruzicka.com/java-cloning-problems/ so cloneable is hardly useful in most cases, if one needs multiple instances of a component you might want to use a factory component instead or the component context.

Reply:
> reusing as much as possible and only overriding the basedir to customize the executionRequest

But that the property of the `ExecutionRequest`. If we are talking about extensions and alike we need a *deep* copy. So for the sake of MultiModuleProject dir I think this is not something we should "clone" for and for the other case we should use a separate instance (as it would require new container).

Reply:
This patch is not about extensions at the moment. Let's stick to what it's supposed to fix now and we can improve later to grow to other cases.

Reply:
Yep that's why I think it should not be needed to clone/copy/... anything in `IMaven` at the moment but purely handle this in the `ExecutionContext`.

Reply:
We already have 1 new API to customize how project execute stuff. We need to be able to leverage it. What you suggest is adding an extra API that does cover only a small part of what the current API -which already exists- can enable. So the proposal is weaker than the current state while adds complexity/API. Multiplying entry points that can achieve the same thing is not really helpful.

Reply:
> We already have 1 new API to customize how project execute stuff.

**Setup** and **Execution** are just two different parts in maven, so if you want to configure the execution it should be local to that execution and not global to the setup phase. and I don't propose any new API here, but cloning is the wrong technique here, it will completely disconnect you from any required OSGi life-cycle. If you still think than my non-new-api and local-injection of that property is not sufficient I would suggest we first solve the problem of multiple (independent) IMaven instances before proceed any further here. It will be required anyways for supporting extensions.